docs/adrs/024-remote-module-invocation.md

```diff
@@ -0,0 +1,280 @@
+```python3
+def init(model_config: RemoteModuleConfig)->ModuleBase:
+    # Construct empty RemoteModule Instance
+    @module(
+        id=model_config.id,
+        name=model_config.name,
+        version=model_config.name,
+        task=[task_tuple[0] for task_tuple in model_config.task_methods]
+    )
+    class _RemoteModelInstance(RemoteModelBase):
+        pass
+    
+    # Add all task methods to the RemoteModel class
+    for task, inference_methods in model_config.task_methods:
+        for method in inference_methods:
+            infer_func = partial(_RemoteModelInstance.remote_method_request, method=method)
+            task_wrapped_func = task.taskmethod(infer_func)
+            setattr(_RemoteModelInstance, method.signature.name, task_wrapped_func)
+    
+    # Add train method to class if one exists
+    if model_config.train_method:
+        train_func = partial(_RemoteModelInstance.remote_method_request, method=model_config.train_method)
+        setattr(_RemoteModelInstance, model_config.train_method.signature.name, train_func)
+    
+    # Return Model Instance
+    return _RemoteModelInstance(model_config.connection, model_config.model_path)    
+    
+
+class RemoteModelBase(ModuleBase):
+   def __init__(self, connection: Dict[str, Any], model_path: str):
+        ...
+   
+   def remote_method_request(self, method: RemoteMethodRpc, *args, **kwargs):
+        # Run the remote invocation using the information defined in the RemoteMethodRpc
+        if self.connection.protocol == "grpc":
+            <run remote `method` using grpc>
+        elif self.connection.protocol == "http":
+            <run remote `method` using http>
+```
+
+## RemoteModuleConfig
+
+```yaml
+class RemoteModuleConfig(ModuleConfig):
+  # Remote runtime information copied from the RemoteModelFinder config
+  connection: Dict[str, Any]
+
+  # Method information
+  # use list and tuples instead of a dictionary to avoid aconfig.Config error
+  task_methods: List[Tuple[type[TaskBase], List[RemoteMethodRpc]]]
+  train_method: RemoteMethodRpc
+  
+  # Source Module Information
+  module_id: str
+  module_name: str
+  model_path: str
+
+@dataclass
+class RemoteMethodRpc:
+    # full signature for this RPC
+    signature: CaikitMethodSignature
+    
+    # Request and response objects for this RPC
+    request_dm_name: str
+    response_dm_name: str
+    
+    # Either the function name of the GRPC Servicer or HTTP endpoint
+    rpc_name: str
+
+    # Only used for infer RPC types
+    input_streaming: bool
+    output_streaming: bool
+```
+
+The RemoteModuleConfig is a custom subclass of ModuleConfig, which contains a description of a Module's
+tasks, inference and train methods, and version information, as well as the connection information of 
+the remote runtime. The combination of the two allows the RemoteModelInitializer to construct a new 
+RemoteModule without having to import `caikit.runtime` or the source model. To help simplify the
```

Comment:
Ah! I forgot that we had moved `CaikitMethodSignature` to `core`. This makes it really nice and clean.